src/cdomain/value/cdomains/int/bitfieldDomain.ml

```diff
@@ -0,0 +1,623 @@
+        else top_of ik in
+      norm ik res
+
+  let rem ik (z1, o1) (z2, o2) = 
+    if o2 = Ints_t.zero then top_of ik else
+    if BArith.is_const (z1, o1) && BArith.is_const (z2, o2) then (
+      let tmp = o1 %: o2 in (!:tmp, tmp)      
+    )
+    else if BArith.is_const (z2, o2) && is_power_of_two o2 then (
+      let mask = Ints_t.sub o2 Ints_t.one in
+      let newz = Ints_t.logor z1 (Ints_t.lognot mask) in
+      let newo = Ints_t.logand o1 mask in
+      norm ik (newz, newo) |> fst
+    )
+    else top_of ik
+
+  let eq ik x y =
+    if (BArith.max ik x) <= (BArith.min ik y) && (BArith.min ik x) >= (BArith.max ik y) then of_bool ik true 
+    else if (BArith.min ik x) > (BArith.max ik y) || (BArith.max ik x) < (BArith.min ik y) then of_bool ik false 
+    else BArith.top_bool
+
+  let ne ik x y = match eq ik x y with
+    | t when t = of_bool ik true -> of_bool ik false
+    | t when t = of_bool ik false -> of_bool ik true
+    | _ -> BArith.top_bool
+
+  let le ik x y = 
+    if (BArith.max ik x) <= (BArith.min ik y) then of_bool ik true 
+    else if (BArith.min ik x) > (BArith.max ik y) then of_bool ik false 
+    else BArith.top_bool
+
+  let ge ik x y = le ik y x
+
+  let lt ik x y = if (BArith.max ik x) < (BArith.min ik y) then of_bool ik true 
+    else if (BArith.min ik x) >= (BArith.max ik y) then of_bool ik false 
+    else BArith.top_bool
+
+  let gt ik x y = lt ik y x
+
+  (* Invariant *)
+
+  let invariant_ikind e ik (z,o) = 
+    let range = range ik (z,o) in
+    IntInvariant.of_interval e ik range
+
+  let starting ?(suppress_ovwarn=false) ik n = 
+    let (min_ik, max_ik) = Size.range ik in
+    of_interval ~suppress_ovwarn ik (n, Ints_t.of_bigint max_ik) 
+
+  let ending ?(suppress_ovwarn=false) ik n =
+    let (min_ik, max_ik) = Size.range ik in
+    of_interval ~suppress_ovwarn ik (Ints_t.of_bigint min_ik, n)
+
+  (* Refinements *)
+
+  let refine_with_congruence ik bf ((cong) : (int_t * int_t ) option) : t =
+    match bf, cong with
+    | (z,o), Some (c, m) when m = Ints_t.zero -> norm ik (!: c, c) |> fst
+    | (z,o), Some (c, m) when is_power_of_two m && m <> Ints_t.one ->
+      let congruenceMask = !: (m -: Ints_t.one) in
+      let congZ = congruenceMask |: !:c in 
+      let congO = congruenceMask |: c in 
+      meet ik (congZ, congO) bf
+    | _ -> norm ik bf |> fst
+
+  let refine_with_interval ik t itv = 
+    match itv with
+    | None -> norm ik t |> fst
+    | Some (l, u) -> meet ik t (of_interval ik (l, u) |> fst)  
+
+  let refine_with_bitfield ik x y = meet ik x y
+
+  let refine_with_excl_list ik t (excl : (int_t list * (int64 * int64)) option) : t = norm ik t |> fst
+
+  let refine_with_incl_list ik t (incl : (int_t list) option) : t =
+    let joined =match incl with
+      | None -> top_of ik
+      | Some ls -> 
+        List.fold_left (fun acc i -> BArith.join acc (BArith.of_int i)) (bot_of ik) ls
+    in
+    meet ik t joined
+
+
+  (* Unit Tests *)
+
+  let arbitrary ik = 
```

Comment:
`arbitrary` is used for unit tests, indeed; still the comment may be a bit misleading, since there are no unit tests at following the comment.